axes_manager getitem iterable

to266 opened the pull request:
Enables `axes_manager` to deal with iterables and axes in `getitem`. Hence `signal.sum(('x', 'y', 'z'))` and similar things are possible.

Main reason for this PR is that when working with large datasets, recreating a new signal for every operation (when multiple are "chained") is very expensive..

as a bonus, things like
```python
s.sum(s.axes_manager.signal_axes + s.axes_manager.navigation_axes[:-1])
```
works as well, if you want to sum over all except the last navigation axis.

Let me know if you have any ideas (if and where) I should mention this in user guide and what else to test

probably solves a part of https://github.com/hyperspy/hyperspy/issues/814, and implements https://github.com/hyperspy/hyperspy/pull/422 to some degree..

to266 commented:
Just to be clear, duplicates are removed, but unique order is preserved:
```python
In [2]: s = hs.signals.Signal(np.random.random((3,2,4,5)))

In [3]: s.axes_manager[-2].name = 'x'   

In [4]: interesting_axes = ('x', 0, -1, s.axes_manager.navigation_axes[-1])

In [5]: interesting_axes
Out[5]: ('x', 0, -1, <x axis, size: 3, index: 0>)

In [6]: axes = s.axes_manager[interesting_axes]

In [7]: axes
Out[7]: 
(<x axis, size: 3, index: 0>,
 <Unnamed 0th axis, size: 4, index: 0>,
 <Unnamed 3rd axis, size: 5>)

In [8]: s.axes_manager['x']
Out[8]: <x axis, size: 3, index: 0>

In [9]: s.axes_manager[0]
Out[9]: <Unnamed 0th axis, size: 4, index: 0>

In [10]: s.axes_manager[-1]
Out[10]: <Unnamed 3rd axis, size: 5>

In [11]: s.axes_manager.navigation_axes[-1]
Out[11]: <x axis, size: 3, index: 0>


```

vidartf commented:
I like the principle a lot! I'll go through the details, but I think this can well replace #422.

vidartf commented on hyperspy/axes.py:
```diff
@@ -504,6 +504,20 @@ def __getitem__(self, y):
+        return_one = False
+        if isinstance(y, basestring) or not np.iterable(y):
+            y = (y,)
+            return_one = True
```
Maybe more elegant to replace `return_one = True` with `return self[(y,)][0]` ? Simplifies the rest of the function as well, but relies on recursion (might be an issue with inheritance).

to266 replied:
Well spotted, will push in a second

vidartf commented on hyperspy/signal.py:
```diff
@@ -3936,12 +3934,15 @@ def _iterate_signal(self):
+        axes = am[axes]
+        if not np.iterable(axes):
+            axes = (axes,)
+        if am.navigation_dimension + am.signal_dimension > len(axes):
```
Implement and use?
```python
def __len__(self):
    return len(self._axes)
```
`am.navigation_dimension + am.signal_dimension` --> `len(am)`

vidartf replied:
Or maybe that can be confused with iteration length (since AxesManager iterates of navigation axes)..

vidartf commented:
As mentioned in #422, the docstring for `axes` can probably be extracted to `hyperspy.docstring.signal.py` similarly to what is done in #424. Other than that I have two questions for discussion:
 1. You suggest renaming `axis` args to `axes`. That will break any existing keyword use. That's OK with me, but it might break users' scripts.
 2. Functions that still only take a single `axis`, should those be extended to also take multiple axes (apply inner code iteratively)?

francisco-dlp commented:
Regarding point 1, I suggest not to rename `axis` -> `axes`. This, in addition of not breaking a lot of code out there, is consistent with numpy's syntax.

vidartf commented:
For (2): `Signal.valuemax` can just wrap `np.amax` which supports tuple `axis` arg. We should also consider renaming it as `amax`, and making `valuemax` a (deprecated?) alias.

vidartf commented:
For (1), @francisco-dlp's point that numpy uses `axis` makes me be +1 on keeping `axis`. That would also make this PR fully compatible with the API, as only protected functions change name.

to266 commented:
I've moved the `axis` docstring out, renamed back to the original name the keyword (even though it never actually was a "keyword"... it's just a positional argument in our case, hence should not matter).

I've also added `signal.amax`, since it does not actually deprecate `valuemax` - they do different things. `indexmax` returns indices of the maximum, `valuemax` returns axes values at those indices, and `amax` - the actual maximum

vidartf commented:
`EDSSpectrum.sum()` currently fails when called with a tuple of axes (trying to plot a spectrum image e.g.). On the line:
```python
self.axes_manager[axis].size
```
Should probably also add a test or modify the existing test to check that the live time calculation is correct also for more than one axis.

to266 commented:
I've fixed the bug, and in general made the calculation more robust (at least I think so..)

As for the (2), I don't think it's worth implementing the functions ourselves. If I really end up needing that, I will at another PR, however here I think just making use of what numpy is capable of is enough

francisco-dlp commented on hyperspy/docstrings/signal.py:
```diff
@@ -0,0 +1,15 @@
+            axis in `axes_manager` or the axis name."""
+MANY_AXIS_PARAMETER = \
+    """: {int | string | axis | tuple}
+            Either one on its own, or many axes in a tuple can be passed. In
```
Shouldn't the default be summing over all navigation axes?

to266 replied:
You think I should add it in this PR? Here I only expanded API without changing behaviour, as that was not the intended purpose of the PR.

francisco-dlp replied:
I agree.

vidartf commented:
Aside from what @francisco-dlp has found / will find, I'm happy with its current state. If you'd like to be the second opinion @francisco-dlp then you can merge it when you think it is OK :)
